File: algorithms/0205.isomorphic-strings/solution.go
func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	m := map[rune]rune{}
	t_chars := []rune(t)

	for i, c := range []rune(s) {
		if _, ok := m[c]; ok {
			if m[c] != t_chars[i] {
				return false
			}
		} else {
			m[c] = t_chars[i]
		}
	}

	values := make(map[rune]bool, len(s))
	for _, val := range m {
		if values[val] {
			return false
		} else {
			values[val] = true
		}
	}

	return true
}

File: algorithms/0205.isomorphic-strings/solution.rb
# @param {String} s
# @param {String} t
# @return {Boolean}
def is_isomorphic(s, t)
  return false unless s.length == t.length

  hash = {}
  t_chars = t.chars

  s.each_char.with_index do |c, i|
    if hash[c]
      return false unless hash[c] == t_chars[i]
    else
      hash[c] = t_chars[i]
    end
  end

  values = hash.values.sort
  values.size == values.uniq.size
end

